### ISSUE TITLE
Supported communication pattern of the DataMan engine

### ISSUE THREAD

**Issue by fengggli** [ornladios/ADIOS2#3068]
Hi!

## Question Summary
The [current documentation](https://adios2.readthedocs.io/en/latest/engines/engines.html#dataman-for-wide-area-network-data-staging) says DataMan engine "is supposed to be used in cases where a few writers send data to a few readers over long distance."
I am trying to understand what cases (NxM, Nx1, 1x1) are supported from the examples.

I have a few questions listed below:
1. I have tried the examples (examples/hello and example/heatTransfer) and can only get them working for the 1x1 case (for the 2.7.1 release), is this expected?
2. The [hello example readme file](https://github.com/ornladios/ADIOS2/blob/master/examples/hello/README.md) mentioned the DataMan transport is "to be deprecated", does it mean it won't be maintained in the future and it's not suggested to use this engine for current projects?
3. I saw the Nx1 mode worked previously in https://github.com/ornladios/ADIOS2/pull/433#issuecomment-362468644 , just wondering are there working examples of such usages in the current source tree?

Thanks!
Feng Li

### heat-transfer

The heat-transfer example seems to have [an outdated XML](https://github.com/ornladios/ADIOS2/blob/master/examples/heatTransfer/heat_dataman.xml) file for DataMan engine.
I modified it a little for the "writer" and "readerInput" and it runs well for the 1x1 case:
```
<io name="writer">
        <engine type="DataMan">

            <parameter key="IPAddress" value="127.0.0.1"/>
            <parameter key="Port" value="13200"/>
            <parameter key="Timeout" value="5"/>
        </engine>
    </io>


    <!--=======================================
           Configuration for the Reader input
        =======================================-->

    <io name="readerInput">
        <engine type="DataMan">
            <parameter key="IPAddress" value="127.0.0.1"/>
            <parameter key="Port" value="13200"/>
            <parameter key="Timeout" value="5"/>
        </engine>
    </io>
```

#### producer
```
heatTransfer$mpirun -n 1 $INST_DIR/heatTransfer_write_adios2  heat_dataman.xml heat.bp 1 1 2 10 10 1
Step 1:
Step 2:
Step 3:
Step 4:
Step 5:
Step 6:
Step 7:
Step 8:
Step 9:
Step 10:
Total runtime = 6.02127s
```

#### consumer
```
heatTransfer$mpirun -n 1 $INST_DIR/heatTransfer_read heat_dataman.xml heat.bp out.bp 1 1
gndx       = 2
gndy       = 10
rank 0 reads 2D slice 2 x 10 from offset (0,0)
Processing step 0
Processing step 1
```

#### Producer:
```
heatTransfer$mpirun -n 2 $INST_DIR/heatTransfer_write_adios2  heat_dataman.xml heat.bp 2 1 2 10 10 1

```
#### Consumer:
```
heatTransfer$mpirun -n 2 $INST_DIR/heatTransfer_read heat_dataman.xml heat.bp out.bp 2 1
[DESKTOP-B450FL:23189] *** Process received signal ***                                                                                                                                                           [DESKTOP-B450FL:23189] Signal: Segmentation fault (11)
[DESKTOP-B450FL:23189] Signal code: Address not mapped (1)
[DESKTOP-B450FL:23189] Failing at address: 0x100000619
[DESKTOP-B450FL:23189] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0)[0x7fb51995a3c0]
[DESKTOP-B450FL:23189] [ 1] /home/fengggli/software/spack/var/spack/environments/adios2-py3/.spack-env/view/lib/../lib/../lib/libzmq.so.5(+0x4f7a4)[0x7fb51885c7a4]
[DESKTOP-B450FL:23189] [ 2] /home/fengggli/software/spack/var/spack/environments/adios2-py3/.spack-env/view/lib/../lib/../lib/libzmq.so.5(+0x6f10f)[0x7fb51887c10f]
[DESKTOP-B450FL:23189] [ 3] /home/fengggli/software/spack/var/spack/environments/adios2-py3/.spack-env/view/lib/../lib/../lib/libzmq.so.5(zmq_recv+0x30)[0x7fb51887cad0]
[DESKTOP-B450FL:23189] [ 4] /home/fengggli/software/spack/var/spack/environments/adios2-py3/.spack-env/view/lib/../lib/libadios2_core.so.2(_ZN6adios23zmq9ZmqPubSub7ReceiveEv+0x31)[0x7fb5193470e1]
[DESKTOP-B450FL:23189] [ 5] /home/fengggli/software/spack/var/spack/environments/adios2-py3/.spack-env/view/lib/../lib/libadios2_core.so.2(_ZN6adios24core6engine13DataManReader15SubscribeThreadERNS_3zmq9ZmqPubSubE+0x5a)[0x7fb51938a23a]
[DESKTOP-B450FL:23189] [ 6] /lib/x86_64-linux-gnu/libstdc++.so.6(+0xd6de4)[0x7fb519a59de4]
[DESKTOP-B450FL:23189] [ 7] /lib/x86_64-linux-gnu/libpthread.so.0(+0x9609)[0x7fb51994e609]
[DESKTOP-B450FL:23189] [ 8] /lib/x86_64-linux-gnu/libc.so.6(clone+0x43)[0x7fb519875293]
[DESKTOP-B450FL:23189] *** End of error message ***
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 0 on node DESKTOP-B450FL exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```

Note: 2x2 cases work fine if replacing "heat_dataman.xml" with "heat_sst_bp.xml"

### environment
ADIOS2 is installed in a ubuntu20.04 environment with spack:
```
heatTransfer$spack find
==> In environment adios2-py3
==> Root specs
adios2 +dataman+python

==> 56 installed packages
-- linux-ubuntu20.04-zen2 / gcc@9.3.0 ---------------------------
adios2@2.7.1         cmake@3.22.2        gettext@0.21          libffi@3.3         libsodium@1.0.18  ncurses@6.2      perl@5.34.0        py-setuptools@59.4.0  sz@2.1.12               zlib@1.2.11
autoconf@2.69        diffutils@3.8       hwloc@2.7.0           libiconv@1.16      libtool@2.4.6     numactl@2.0.14   pkgconf@1.8.0      py-wheel@0.37.0       tar@1.34                zstd@1.5.0
automake@1.16.5      docbook-xml@4.5     libbsd@0.11.3         libmd@1.0.3        libxml2@2.9.12    openblas@0.3.19  py-cython@0.29.24  python@3.9.10         util-linux-uuid@2.36.2
berkeley-db@18.1.40  docbook-xsl@1.79.2  libedit@3.1-20210216  libpciaccess@0.16  libzmq@4.3.4      openmpi@4.1.2    py-mpi4py@3.1.2    readline@8.1          util-macros@1.19.3
bzip2@1.0.8          expat@2.4.4         libevent@2.1.12       libpng@1.6.37      lz4@1.9.3         openssh@8.8p1    py-numpy@1.22.2    snappy@1.1.8          xz@5.2.5
c-blosc@1.21.0       gdbm@1.19           libfabric@1.14.0      libsigsegv@2.13    m4@1.4.19         openssl@1.1.1m   py-pip@21.3.1      sqlite@3.37.1         zfp@0.5.5
```

PS: I remember dealing with different versions of software and off-tree dependencies in  ADIOS1 has been a pain, and supporting spack makes it so easy in adios2. Really appreciate this great work!

**Comment by JasonRuonanWang** [comment]
Thanks for pointing out the issues. I will modify the doc later to avoid confusions. 

DataMan is only designed for data transfer over wide area networks. In this case, network latency is usually at ~10ms to ~100ms level, which makes it extremely inefficient for NxM cases. We used to support NxM at a small scale but later we decided to focus on 1x1 cases for best performance. This is also inline with common topologies we have seen. People don't usually use supercomputer nodes where they run large-scale jobs for long distance data transfers. Instead, these data transfers usually happen on data transfer nodes (DTN) which usually do not support multi-node MPI jobs anyway. 

For any multi-node local jobs, SST and SSC are the right engines to use. 

I didn't know about the deprecation notice of DataMan, and I don't understand why it's there. The only reasonable explanation to me is that the hello world example is to be deprecated, but absolutely not the DataMan engine itself.
